File: engine/helengine.core/shaders/ShaderModuleDefinition.cs
namespace helengine {
    /// <summary>
    /// Groups shader programs and binaries for a single shader module.
    /// </summary>
    public class ShaderModuleDefinition {
        /// <summary>
        /// Stores the shader program definitions for this module.
        /// </summary>
        readonly ShaderProgramDefinition[] programs;

        /// <summary>
        /// Stores the compiled binaries for this module.
        /// </summary>
        readonly ShaderProgramBinary[] binaries;

        /// <summary>
        /// Initializes a new shader module definition.
        /// </summary>
        /// <param name="name">Module name.</param>
        /// <param name="programs">Shader program definitions.</param>
        /// <param name="binaries">Compiled binaries for all targets and variants.</param>
        public ShaderModuleDefinition(
            string name,
            ShaderProgramDefinition[] programs,
            ShaderProgramBinary[] binaries) {
            if (string.IsNullOrWhiteSpace(name)) {
                throw new ArgumentException("Module name must be provided.", nameof(name));
            }

            if (programs == null) {
                throw new ArgumentNullException(nameof(programs));
            }

            if (programs.Length == 0) {
                throw new ArgumentException("At least one program definition is required.", nameof(programs));
            }

            if (binaries == null) {
                throw new ArgumentNullException(nameof(binaries));
            }

            Name = name;
            this.programs = programs;
            this.binaries = binaries;
        }

        /// <summary>
        /// Gets the module name.
        /// </summary>
        public string Name { get; }

        /// <summary>
        /// Gets the shader program definitions.
        /// </summary>
        public IReadOnlyList<ShaderProgramDefinition> Programs {
            get {
                return programs;
            }
        }

        /// <summary>
        /// Gets the compiled binaries for the module.
        /// </summary>
        public IReadOnlyList<ShaderProgramBinary> Binaries {
            get {
                return binaries;
            }
        }

        /// <summary>
        /// Returns the program definition matching the requested name.
        /// </summary>
        /// <param name="programName">Program name to locate.</param>
        /// <returns>Matching program definition.</returns>
        public ShaderProgramDefinition GetProgram(string programName) {
            if (string.IsNullOrWhiteSpace(programName)) {
                throw new ArgumentException("Program name must be provided.", nameof(programName));
            }

            for (int i = 0; i < programs.Length; i++) {
                ShaderProgramDefinition program = programs[i];
                if (string.Equals(program.Name, programName, StringComparison.Ordinal)) {
                    return program;
                }
            }

            throw new InvalidOperationException("No shader program was found for the requested name.");
        }

        /// <summary>
        /// Attempts to locate a program definition by name.
        /// </summary>
        /// <param name="programName">Program name to locate.</param>
        /// <param name="program">Matching program definition when found.</param>
        /// <returns>True when the program is found.</returns>
        public bool TryGetProgram(string programName, out ShaderProgramDefinition program) {
            if (string.IsNullOrWhiteSpace(programName)) {
                throw new ArgumentException("Program name must be provided.", nameof(programName));
            }

            for (int i = 0; i < programs.Length; i++) {
                ShaderProgramDefinition candidate = programs[i];
                if (string.Equals(candidate.Name, programName, StringComparison.Ordinal)) {
                    program = candidate;
                    return true;
                }
            }

            program = null;
            return false;
        }

        /// <summary>
        /// Returns the compiled binary for the requested program, target, and variant.
        /// </summary>
        /// <param name="programName">Program name to locate.</param>
        /// <param name="target">Target backend identifier.</param>
        /// <param name="variant">Variant name.</param>
        /// <returns>Matching binary descriptor.</returns>
        public ShaderProgramBinary GetBinary(string programName, string target, string variant) {
            ShaderProgramBinary binary;
            if (TryGetBinary(programName, target, variant, out binary)) {
                return binary;
            }

            throw new InvalidOperationException("No compiled shader binary was found for the requested selection.");
        }

        /// <summary>
        /// Attempts to locate a compiled binary for the requested program, target, and variant.
        /// </summary>
        /// <param name="programName">Program name to locate.</param>
        /// <param name="target">Target backend identifier.</param>
        /// <param name="variant">Variant name.</param>
        /// <param name="binary">Matching binary descriptor when found.</param>
        /// <returns>True when a matching binary is found.</returns>
        public bool TryGetBinary(string programName, string target, string variant, out ShaderProgramBinary binary) {
            if (string.IsNullOrWhiteSpace(programName)) {
                throw new ArgumentException("Program name must be provided.", nameof(programName));
            }

            if (string.IsNullOrWhiteSpace(target)) {
                throw new ArgumentException("Target must be provided.", nameof(target));
            }

            if (string.IsNullOrWhiteSpace(variant)) {
                throw new ArgumentException("Variant must be provided.", nameof(variant));
            }

            for (int i = 0; i < binaries.Length; i++) {
                ShaderProgramBinary candidate = binaries[i];
                if (!string.Equals(candidate.ProgramName, programName, StringComparison.Ordinal)) {
                    continue;
                }

                if (!string.Equals(candidate.Target, target, StringComparison.OrdinalIgnoreCase)) {
                    continue;
                }

                if (string.Equals(candidate.Variant, variant, StringComparison.Ordinal)) {
                    binary = candidate;
                    return true;
                }
            }

            binary = null;
            return false;
        }
    }
}

File: tmp/helengine-core-cpp-regenerated/ShaderModuleDefinition.cpp
#ifdef DrawText
#undef DrawText
#endif
#include "ShaderModuleDefinition.hpp"
#include "runtime/native_string.hpp"
#include "runtime/native_exceptions.hpp"
#include "ShaderProgramBinary.hpp"
#include "ShaderProgramDefinition.hpp"
#include "runtime/array.hpp"
#include "runtime/finally.hpp"
#include "runtime/native_cast.hpp"
#include "runtime/native_datetime.hpp"
#include "runtime/native_dictionary.hpp"
#include "runtime/native_disposable.hpp"
#include "runtime/native_enum.hpp"
#include "runtime/native_event.hpp"
#include "runtime/native_exceptions.hpp"
#include "runtime/native_list.hpp"
#include "runtime/native_nullable.hpp"
#include "runtime/native_span.hpp"
#include "runtime/native_stack.hpp"
#include "runtime/native_string.hpp"
#include "runtime/native_tuple.hpp"
#include "runtime/native_type.hpp"
#include "system/app_context.hpp"
#include "system/bit_converter.hpp"
#include "system/diagnostics/debug.hpp"
#include "system/diagnostics/stopwatch.hpp"
#include "system/guid.hpp"
#include "system/io/file-stream.hpp"
#include "system/io/file.hpp"
#include "system/io/memory-stream.hpp"
#include "system/io/path.hpp"
#include "system/io/stream-reader.hpp"
#include "system/io/stream.hpp"
#include "system/io/string-reader.hpp"
#include "system/math.hpp"
#include "system/number.hpp"
#include "system/string_comparer.hpp"
#include "system/text/encoding.hpp"
#include "system/text/regular_expressions/regex.hpp"
#include "system/text/string-builder.hpp"

List<::ShaderProgramBinary*>* ShaderModuleDefinition::get_Binaries()
{
return new List<ShaderProgramBinary*>(this->binaries);}

std::string ShaderModuleDefinition::get_Name()
{
return this->Name;
}

List<::ShaderProgramDefinition*>* ShaderModuleDefinition::get_Programs()
{
return new List<ShaderProgramDefinition*>(this->programs);}

::ShaderProgramBinary* ShaderModuleDefinition::GetBinary(std::string programName, std::string target, std::string variant)
{
::ShaderProgramBinary *binary;
    if (this->TryGetBinary(programName, target, variant, binary))
    {
return binary;    }
throw new InvalidOperationException("No compiled shader binary was found for the requested selection.");
}

::ShaderProgramDefinition* ShaderModuleDefinition::GetProgram(std::string programName)
{
    if (String::IsNullOrWhiteSpace(programName))
    {
throw ([&]() {
auto __ctor_arg_00000136 = "Program name must be provided.";
auto __ctor_arg_00000137 = "programName";
return new ArgumentException(__ctor_arg_00000136, __ctor_arg_00000137);
})();
    }
for (int32_t i = 0; i < this->programs->Length; i++) {
::ShaderProgramDefinition *program = (*this->programs)[i];
    if (String::Equals(program->get_Name(), programName, StringComparison::Ordinal))
    {
return program;    }
}
throw new InvalidOperationException("No shader program was found for the requested name.");
}

ShaderModuleDefinition::ShaderModuleDefinition(std::string name, Array<::ShaderProgramDefinition*>* programs, Array<::ShaderProgramBinary*>* binaries) : Name(), binaries(), programs()
{
    if (String::IsNullOrWhiteSpace(name))
    {
throw ([&]() {
auto __ctor_arg_00000138 = "Module name must be provided.";
auto __ctor_arg_00000139 = "name";
return new ArgumentException(__ctor_arg_00000138, __ctor_arg_00000139);
})();
    }
    if (programs == nullptr)
    {
throw new ArgumentNullException("programs");
    }
    if (programs->Length == 0)
    {
throw ([&]() {
auto __ctor_arg_0000013A = "At least one program definition is required.";
auto __ctor_arg_0000013B = "programs";
return new ArgumentException(__ctor_arg_0000013A, __ctor_arg_0000013B);
})();
    }
    if (binaries == nullptr)
    {
throw new ArgumentNullException("binaries");
    }
this->Name = name;
this->programs = programs;
this->binaries = binaries;
}

bool ShaderModuleDefinition::TryGetBinary(std::string programName, std::string target, std::string variant, ::ShaderProgramBinary*& binary)
{
    if (String::IsNullOrWhiteSpace(programName))
    {
throw ([&]() {
auto __ctor_arg_0000013C = "Program name must be provided.";
auto __ctor_arg_0000013D = "programName";
return new ArgumentException(__ctor_arg_0000013C, __ctor_arg_0000013D);
})();
    }
    if (String::IsNullOrWhiteSpace(target))
    {
throw ([&]() {
auto __ctor_arg_0000013E = "Target must be provided.";
auto __ctor_arg_0000013F = "target";
return new ArgumentException(__ctor_arg_0000013E, __ctor_arg_0000013F);
})();
    }
    if (String::IsNullOrWhiteSpace(variant))
    {
throw ([&]() {
auto __ctor_arg_00000140 = "Variant must be provided.";
auto __ctor_arg_00000141 = "variant";
return new ArgumentException(__ctor_arg_00000140, __ctor_arg_00000141);
})();
    }
for (int32_t i = 0; i < this->binaries->Length; i++) {
::ShaderProgramBinary *candidate = (*this->binaries)[i];
    if (!String::Equals(candidate->get_ProgramName(), programName, StringComparison::Ordinal))
    {
continue;
    }
    if (!String::Equals(candidate->get_Target(), target, StringComparison::OrdinalIgnoreCase))
    {
continue;
    }
    if (String::Equals(candidate->get_Variant(), variant, StringComparison::Ordinal))
    {
binary = candidate;
return true;    }
}
binary = nullptr;
return false;}

bool ShaderModuleDefinition::TryGetProgram(std::string programName, ::ShaderProgramDefinition*& program)
{
    if (String::IsNullOrWhiteSpace(programName))
    {
throw ([&]() {
auto __ctor_arg_00000142 = "Program name must be provided.";
auto __ctor_arg_00000143 = "programName";
return new ArgumentException(__ctor_arg_00000142, __ctor_arg_00000143);
})();
    }
for (int32_t i = 0; i < this->programs->Length; i++) {
::ShaderProgramDefinition *candidate = (*this->programs)[i];
    if (String::Equals(candidate->get_Name(), programName, StringComparison::Ordinal))
    {
program = candidate;
return true;    }
}
program = nullptr;
return false;}

